### DISCUSSION TITLE
Welcome to the AWS SDK for Swift Discussions!

### DISCUSSION THREAD

**Discussion by kneekey23** [awslabs/aws-sdk-swift#270] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by rromanchuk** [comment]
Thank you for doing this!  Very motivated to contribute, especially with reaching parity of the well scoped/constrained obc-c wrappers (S3/Dynamo specifically). The objC  client-direct-to-aws-service-via-cognito-(identity pool) really shines in novel ways in a mobile context. It was starting to feel like the obj-c sdk was just a request generator for amplify.  My number one request would be to keep this namespaced away to make www useful again. "aws ios sdk"  became synonymous with amplify where low-level topics were entirely crowded out.

**Reply by kneekey23** [reply]
we won't have any of those higher level wrappers in this sdk, it will be a purely generated sdk. in the distant future we could expand to that but we leave those higher level wrappers to Amplify. This is an AWS SDK now and the iOS SDK is built and maintained by the mobile org. This SDK is now built and maintained by SDK & Tools at AWS just like all the other official AWS SDKs. :)

**Reply by kneekey23** [reply]
Amplify will eventually migrate their plugins to depend on this SDK as well.

**Reply by kneekey23** [reply]
Hopefully this answers your question/request? In fact no need for a request because you can consider it done. ha

**Comment by trvd1707** [comment]
I'm having a problem trying to add this swift package into my project using Xcode 13.1 It gives me an error: AWS-sdk-swift could not be resolved. InvalidManifestFormat.

I'm able to add the Objc framework as a swift package, but then I get ambiguous definition when I try to use uploadData method

**Reply by kneekey23** [reply]
can you paste your manifest here possibly?

**Comment by trvd1707** [comment]
I don't have the manifest. I just provided this repo's to Swift package
manager and tried to import the package. What I found out is that the
manifest in the repo at the master branch has this problem, but if I import
the tag 0.1.0, it is able to load the package without problem. Now I'm
finding out that the interface is quite different than the ObjC one.

**Reply by kneekey23** [reply]
you can't use the sdk without a Package manifest? You must have a Package.swift file or are you trying to use it in an iOS app cuz then you need to go through Xcode to add a package and yes you need to use the tagged version. There is no version of the sdk in the main branch. This SDK is a brand new Swift SDK and has no relation whatsoever to the ObjC SDK FYI. You can documentation for this SDK here: https://docs.aws.amazon.com/sdk-for-swift/latest/developer-guide/home.html

**Comment by trvd1707** [comment]
Well, I'm using xcode swift package manager. When I select Add package in
xcode, it asks for the URL where the package is and it shows me this
package. If I don't specify the version, it grabs from the master branch.
When parsing the master branch, it gives me this error. If I specify the
specific tag 0.1.0, Swift package manager is able to parse the package and
provides me the list of products in it and allows me to choose which
targets will be dependent on which products of the package.  In Xcode 13,
the package manager builds the package.swift for me and it is hidden from
the developer.

Now I'm struggling to re-write the code that was written using the ObjC to
use the swift package. Everything was prefixed by AWS in the Objc package,
but certain classes and methods available in the objc package have
different names.

**Reply by kneekey23** [reply]
Have you checked out the documentation I linked? Everything you're looking for should be in there including the set up for an iOS application. And yes you always need a tagged version, there is no SDK in the main branch.

**Reply by trvd1707** [reply]
I'm still trying to understand the documentation, but without luck so far.
My code has to run from iOS 13 and up and this swift package seems to have a lot of functions that require iOS15.

Even restricting my code to iOS 15, I couldn't get even the first three lines to be be converted to the new package:

let credentialsProvider = AWSCognitoIdentityProvider(regionType:.USEast1, identityPoolId:"<myIdentityPoolId>") 
let configuration = AWSServiceConfiguration(region:.USEast1, credentialsProvider:credentialsProvider)  
AWSServiceManager.default().defaultServiceConfiguration = configuration

The swift package has a CognitoIdentityProviderClient, but its initializer does not take a the identityPoolId

**Comment by trvd1707** [comment]
Thanks for the documentation link. I’m no aws expert and I’ve been working with sample code that was based on the ObjC framework. Steep learning curve…

— Tt

**Comment by trvd1707** [comment]
My question is why the swift package had to use completely different names and even method signatures than of the the ObjC version? And why the decision of using iOS15 only?

**Comment by kneekey23** [comment]
Hi @trvd1707,

First off sorry for the delay, I have been out on vacation. Several questions here to respond to:
1. First off this new SDK is the low level code generated SDK generated purely from the service APIs. The old objc SDK is both a low level SDK and a mid level SDK with some hand written code on top to abstract the low level SDK away from you in certain cases, Cognito being one of them. This is why the interfaces don't look the same to you. But under the hood, it's the same I promise. This SDK will only ever remain a low level SDK. Amplify iOS will become that new abstraction around this new SDK to help you make calls from mobile apps easier. It seems like you might want to look into migrating to Amplify not the low level SDK directly as you are correct in that it will be a steep learning curve. https://docs.amplify.aws/start/q/integration/ios/
2. The SDK will support iOS 13 and above. The async functions right now don't support that until this PR is merged: https://github.com/awslabs/aws-sdk-swift/pull/497 due to Apple only recently backporting support for the concurrency primitives. There are completion handler APIs right now that the SDK offers in order to use an iOS platform prior to 15 but I discourage using them as they will be removed when that PR is merged. 
3. Let me know if I can answer any other questions.

**Comment by trvd1707** [comment]
Thank you for the clarification.
Yes, I figured out that it would be a while until I could get the swift
package working. Another challenge for me was to combine swift
packages with cocoapods in a single project. You see, I'm  using the AWS
SDK inside a test target, not in the main app. In the end, I switched to
using the Objc pod and it is working for what I need now. Thank you for
your help.

The challenge of using async objects in a test target is that it doesn't
come with runloop that is available inside the regular app, so I have to
create expectations, add a loop to wait for the expectation to be set and
use the closures to set the expectations accordingly, to prevent the test
from finishing before the AWS calls have a chance to call the closures.

I'll take a look at amplify.
